model.py

```diff
@@ -3461,78 +3399,131 @@ def all_descriptions(self):
         # actually changed.
         changed = False
         edition_metadata_changed = False
         classification_changed = False
 
-        primary_edition = self.primary_edition
-        summary = self.summary
-        summary_text = self.summary_text
-        quality = self.quality
+        old_primary_edition = self.primary_edition
+        new_primary_edition = None
+
+        # an open access pool may be "superceded", if there's a better-quality 
+        # open-access pool available
+        self.mark_licensepools_as_superceded()
+
+        for pool in self.license_pools:
+            # a superceded pool's composite edition is not good enough
+            # Note:  making the assumption here that we won't have a situation 
+            # where we marked all of the work's pools as superceded or suppressed. 
+            if pool.superceded or pool.suppressed:
+                continue
+
+            edition_metadata_changed = (
+                edition_metadata_changed or
+                pool.set_presentation_edition(policy)
+            )
+            potential_primary_edition = pool.presentation_edition
+
+            # We currently have no real way to choose between
+            # competing primary editions. But it doesn't matter much
+            # because in the current system there should never be more
+            # than one non-superceded license pool per Work.
+            #
+            # So basically we pick the first available edition and
+            # make it the primary.
+            if (not new_primary_edition
+                or (potential_primary_edition is old_primary_edition and old_primary_edition)):
+                # We would prefer not to change the Work's primary
+                # edition unnecessarily, so if the current primary
+                # edition is still an option, choose it.
+                new_primary_edition = potential_primary_edition
+
+        if new_primary_edition:
+            self.primary_edition = new_primary_edition
+
+        # go through the loser pools, and tell them they lost
+        for pool in self.license_pools:
+            if pool.presentation_edition is self.primary_edition:
+                # make sure edition knows it's primary
+                pool.presentation_edition.is_primary_for_work = True
+            else:
+                pool.mark_edition_primarity(primary_for_work_edition=None)
+
 
+        # Note: policy.choose_edition is true by default.
+        # If we don't have a self.primary_edition by now, we won't get it in 
+        # self.set_primary_edition().  That method will just set all the editions' 
+        # is_primary_for_work attributes to False.
+        policy = policy or PresentationCalculationPolicy()
         if policy.choose_edition or not self.primary_edition:
             self.set_primary_edition()
             WorkCoverageRecord.add_for(
                 self, operation=WorkCoverageRecord.CHOOSE_EDITION_OPERATION
             )
```

Comment:
there's already a coverage record added in set_primary_edition, so now there are two